contracts/extensions/version-control/ConditionalVersionControl.sol

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { ERC1967Upgrade } from "@openzeppelin/contracts/proxy/ERC1967/ERC1967Upgrade.sol";
+import { IConditionalVersionControl } from "../../interfaces/version-control/IConditionalVersionControl.sol";
+import { ErrorHandler } from "../../libraries/ErrorHandler.sol";
+import { AddressArrayUtils } from "../../libraries/AddressArrayUtils.sol";
+import { ErrOnlySelfCall } from "../../utils/CommonErrors.sol";
+
+/**
+ * @title ConditionalVersionControl
+ * @dev A contract that allows conditional version control of contract implementations.
+ */
+abstract contract ConditionalVersionControl is IConditionalVersionControl, ERC1967Upgrade {
+  using ErrorHandler for bool;
+  using AddressArrayUtils for address[];
+
+  address public immutable proxyStorage;
+  address public immutable newVersion;
+  address public immutable currentVersion;
```

Comment:
I think we should name these immutable vars with _ prefix  and provided with getters.

[This LoC](https://github.com/axieinfinity/ronin-dpos-contracts/pull/254/files#diff-7d433e8d769cb29bcfa1b2c989ef7a01bf4022837d203543c26f34b658b2704aR48) makes me confused for a while.